<launch>

  <!-- Launch realsense camera -->
  <include file="$(find realsense2_camera)/launch/rs_d455_camera_dc.launch"/>

  <node pkg="tf" type="static_transform_publisher" name="camera_to_base_link" args="0 0 0 0 0 0 1 /base_link /d455_link 1000" /> 
  
  <arg name="use_imu_filter"         default="true"/>
  <arg name="raw_imu_topic"          default="/d455/imu"/>
  <arg name="imu_topic"              default="/imu/data"/>
  <arg name="raw_pointcloud_topic"   default="/d455/depth/color/points"/>

  <arg name="camera_height"          default="1.0"/>
  <arg name="camera_frame_id"        default="d455_link"/>
  <arg name="base_frame_id"          default="base_link"/>
  <arg name="odom_frame_id"          default="odom"/>
  
  <arg name="max_sensing_range"      default="7.0"/>
  <arg name="grid_unit_size"         default="0.15"/>
  <arg name="grid_length"            default="15.0"/>
  <arg name="grid_width"             default="15.0"/>
  <arg name="grid_height"            default="1.0"/>
  
  <param name="/use_sim_time"        value="false"/>
  
 <group if="$(arg use_imu_filter)">
  <!-- Launch the Complementary filter node -->
  <node pkg="imu_complementary_filter" type="complementary_filter_node" name="complementary_filter_gain_node" output="screen">
    <remap from="input_raw_imu"                                  to = "$(arg raw_imu_topic)" />
    <param name="do_bias_estimation"        type = "bool"        value = "true"/>
    <param name="do_adaptive_gain"          type = "bool"        value = "true"/>
    <param name="use_mag"                   type = "bool"        value = "false"/>
    <param name="gain_acc"                  type = "double"      value = "0.01"/>
    <param name="gain_mag"                  type = "double"      value = "0.01"/>
  </node>
 </group>
  
  <!-- Launch the ground plane segmentation node -->
  <node pkg="pc_gps" type="pc_gps" name="pc_gps" output="screen">
    <remap from="imu/data"                                       to = "$(arg imu_topic)" />
    <remap from="cloud_in"                                       to = "$(arg raw_pointcloud_topic)" />
    <param name="sensor_height"             type = "double"      value = "$(arg camera_height)"/>
    <param name="num_iter"                  type = "int"         value = "5"/>
    <param name="num_lpr"                   type = "double"      value = "0.7"/>
    <param name="th_seeds"                  type = "double"      value = "0.05"/><!--0.5-->
    <param name="th_dist"                   type = "double"      value = "0.03"/><!--0.1-->
    <param name="block_thres_"              type = "double"      value = "0.3"/>
    <param name="th_ceil_"                  type = "double"      value = "$(arg grid_height)"/>
    <param name="th_box_"                   type = "double"      value = "$(arg max_sensing_range)"/>
    <param name="map_unit_size_"            type = "double"      value = "$(arg grid_unit_size)"/>
    <param name="radius_search_"            type = "double"      value = "0.5"/> <!--15-->
    <param name="in_radius_"                type = "int"         value = "50"/> <!--25-->
    <param name="std_th_"                   type = "double"      value = "0.1"/>
    <param name="mean_k_"                   type = "int"         value = "-15"/>
    <param name="alpha"                     type = "double"      value = "0.1"/>
    <param name="SVD_refinement"            type = "bool"        value = "true"/>
    <param name="detect_neg_obstacle"       type = "bool"        value = "false"/>
    <param name="debug"                     type = "bool"        value = "false"/>
    <param name="timer"                     type = "bool"        value = "false"/>
  </node>
  

  <!-- Launch RViz with the demo configuration -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find local_grid_map)/rviz/data_collection2.rviz" /> 

  
</launch>
